engine/src/main/java/org/pentaho/di/job/entries/setvariables/JobEntrySetVariables.java

```diff
@@ -0,0 +1,518 @@
+          }
+        }
+      }
+
+      for ( int i = 0; i < variables.size(); i++ ) {
+        String varname = variables.get( i );
+        String value = variableValues.get( i );
+        int type = variableTypes.get( i );
+
+        if ( replaceVars ) {
+          varname = environmentSubstitute( varname );
+          value = environmentSubstitute( value );
+        }
+
+        // OK, where do we set this value...
+        switch ( type ) {
+          case VARIABLE_TYPE_JVM:
+            if ( value != null ) {
+              System.setProperty( varname, value );
+            } else {
+              System.clearProperty( varname ); // PDI-17536
+            }
+            setVariable( varname, value );
+            Job parentJobTraverse = parentJob;
+            while ( parentJobTraverse != null ) {
+              parentJobTraverse.setVariable( varname, value );
+              parentJobTraverse = parentJobTraverse.getParentJob();
+            }
+            break;
+
+          case VARIABLE_TYPE_ROOT_JOB:
+            // set variable in this job entry
+            setVariable( varname, value );
+            Job rootJob = parentJob;
+            while ( rootJob != null ) {
+              rootJob.setVariable( varname, value );
+              rootJob = rootJob.getParentJob();
+            }
+            break;
+
+          case VARIABLE_TYPE_CURRENT_JOB:
+            setVariable( varname, value );
+
+            if ( parentJob != null ) {
+              String parameterValue = parentJob.getParameterValue( varname );
+              // if not a parameter, set the value
+              if ( parameterValue == null  ) {
+                setEntryStepSetVariable( varname, value );
+              } else {
+                //if parameter, save the initial parameter value for use in reset/clear variables in future calls
+                if ( parameterValue != null && parameterValue != value && !entryStepSetVariablesMap.containsKey( varname ) ) {
+                  setEntryStepSetVariable( varname, parameterValue );
+                }
+              }
+              parentJob.setVariable( varname, value );
+
+            } else {
+              throw new KettleJobException( BaseMessages.getString(
+                PKG, "JobEntrySetVariables.Error.UnableSetVariableCurrentJob", varname ) );
+            }
+            break;
+
+          case VARIABLE_TYPE_PARENT_JOB:
+            setVariable( varname, value );
+
+            if ( parentJob != null ) {
+              parentJob.setVariable( varname, value );
+              Job gpJob = parentJob.getParentJob();
+              if ( gpJob != null ) {
+                gpJob.setVariable( varname, value );
+              } else {
+                throw new KettleJobException( BaseMessages.getString(
+                  PKG, "JobEntrySetVariables.Error.UnableSetVariableParentJob", varname ) );
+              }
+            } else {
+              throw new KettleJobException( BaseMessages.getString(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [KettleJobException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [365].  Consider adding a catch clause of this specific exception in the outer try statement.